packages/tldraw/src/lib/ui/components/TldrawUiLayout/TldrawUiLayout.tsx

```diff
@@ -0,0 +1,228 @@
+import { clamp } from '@tldraw/editor'
+import classNames from 'classnames'
+import { useCallback, useLayoutEffect, useRef } from 'react'
+import { PORTRAIT_BREAKPOINT } from '../../constants'
+import { useBreakpoint } from '../../context/breakpoints'
+
+/** @public */
+export type TldrawUiLayoutHorizontalPosition = 'left' | 'center' | 'right'
+
+/** @public */
+export type TldrawUiLayoutVerticalPosition = 'top' | 'center' | 'bottom'
+
+/** @public */
+export interface TldrawUiLayoutProps {
+	topLeft?: React.ReactNode
+	topCenter?: React.ReactNode
+	topRight?: React.ReactNode
+	centerLeft?: React.ReactNode
+	centerRight?: React.ReactNode
+	bottomLeft?: React.ReactNode
+	bottomCenter?: React.ReactNode
+	bottomRight?: React.ReactNode
+	spacingPx?: number
+	squishTop?: TldrawUiLayoutHorizontalPosition
+	squishBottom?: TldrawUiLayoutHorizontalPosition
+	squishLeft?: TldrawUiLayoutVerticalPosition
+	squishRight?: TldrawUiLayoutVerticalPosition
+	children?: React.ReactNode
+}
+
+/** @public @react */
+export function TldrawUiLayout({
+	topLeft,
+	topCenter,
+	topRight,
+	centerLeft,
+	centerRight,
+	bottomLeft,
+	bottomCenter,
+	bottomRight,
+	spacingPx = 8,
+	squishTop = 'center',
+	squishBottom = 'center',
+	squishLeft = 'center',
+	squishRight = 'top',
+	children,
+}: TldrawUiLayoutProps) {
+	const breakpoint = useBreakpoint()
+
+	const containerRef = useRef<HTMLDivElement>(null)
+	const topLeftRef = useRef<HTMLDivElement>(null)
+	const topCenterRef = useRef<HTMLDivElement>(null)
+	const topRightRef = useRef<HTMLDivElement>(null)
+	const centerLeftRef = useRef<HTMLDivElement>(null)
+	const centerRightRef = useRef<HTMLDivElement>(null)
+	const bottomLeftRef = useRef<HTMLDivElement>(null)
+	const bottomCenterRef = useRef<HTMLDivElement>(null)
+	const bottomRightRef = useRef<HTMLDivElement>(null)
+
+	const updateLayout = useCallback(() => {
+		const container = containerRef.current!
+		const topLeft = topLeftRef.current!
+		const topCenter = topCenterRef.current!
+		const topRight = topRightRef.current!
+		const centerLeft = centerLeftRef.current!
+		const centerRight = centerRightRef.current!
+		const bottomLeft = bottomLeftRef.current!
+		const bottomCenter = bottomCenterRef.current!
+		const bottomRight = bottomRightRef.current!
+
+		const elements = [
+			topLeft,
+			topCenter,
+			topRight,
+			centerLeft,
+			centerRight,
+			bottomLeft,
+			bottomCenter,
+			bottomRight,
+		]
+		const rects = elements.map((element) => element.getBoundingClientRect())
+		const containerRect = container.getBoundingClientRect()
+
+		// we use this weird set up with indexes so that we can do all the measurements in one go,
+		// then apply all the changes in one go. This prevents layout thrashing.
+		const info = { spacingPx, containerRect, elements, rects }
+
+		layout('x', [0, 1, 2], squishIndexes[squishTop], info)
```

Comment:
i might be annoying and say the readability of this is a little tough :) especially since lots of things get passed into the `info` grabbag.

while more verbose, we might consider something like:
```
const elements = {
  [LAYOUT.TOP_LEFT]: topLeft,
  [LAYOUT.TOP_CENTER]: topCenter,
  ...
}
...
const topLayoutItem = [
  LAYOUT.TOP_LEFT,
  LAYOUT.TOP_CENTER,
  LAYOUT.TOP_RIGHT
][squishIndexes[squishTop]]
layout('x', elements[topLayoutItem], info)
```
besides, `squishIndex` only gets used in the `layout` function to index it things so you might as well just index it here. feels like `layout` shouldn't know about this bit.